File: src/main/scala/com/zcunsoft/clklog/analysis/utils/ClickHouseUtil.scala
package com.zcunsoft.clklog.analysis.utils

import ru.yandex.clickhouse.BalancedClickhouseDataSource
import ru.yandex.clickhouse.settings.ClickHouseProperties

import java.sql.Connection

object ClickHouseUtil {

  var conn: Connection = _

  def getConn(ip_ports: String, database: String, userName: String, password: String): Connection = {
    try {
      if (conn == null) {
        val url = if (database.nonEmpty) {
          s"jdbc:clickhouse://$ip_ports/$database"
        } else s"jdbc:clickhouse://$ip_ports/default"

        val ckProperties = new ClickHouseProperties()
        ckProperties.setUser(userName)
        ckProperties.setPassword(password)

        val balanced = new BalancedClickhouseDataSource(url, ckProperties)
        //对每个host进行ping操作, 排除不可用的连接//对每个host进行ping操作, 排除不可用的连接
        balanced.actualize
        conn = balanced.getConnection
        println("clickhouse connection success...")
      }
    } catch {
      case e: Throwable =>
        throw new Exception("clickhouse connection fail,result : " + e.toString)
    }
    conn
  }

  def close() = {
    if (conn != null) conn.close()
  }


}

File: src/main/java/com/zcunsoft/clklog/analysis/utils/ClickHouseUtil.java
package com.zcunsoft.clklog.analysis.utils;

import com.clickhouse.jdbc.ClickHouseDriver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.Connection;
import java.sql.Driver;
import java.sql.SQLException;
import java.util.Properties;

public class ClickHouseUtil {
  private static Connection connection;

  private static final Logger logger = LoggerFactory.getLogger(ClickHouseUtil.class);

  public static Connection getConn(String host, String database, String userName, String password) {
    try {
      String url = "jdbc:clickhouse://" + host + "/" + database;
      Driver driver = new ClickHouseDriver();
      Properties prop = new Properties();
      prop.put("user", userName);
      prop.put("password", password);
      connection = driver.connect(url, prop);
      logger.info("ck conn got");
    } catch (Exception ex) {
      logger.error("ck conn error ", ex);
    }
    return connection;
  }

  public void close() throws SQLException {
    connection.close();
  }
}

